src/VisualStudio/Core/Def/Implementation/TodoComments/VisualStudioTodoCommentsService.cs

```diff
@@ -0,0 +1,195 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Common;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.TodoComments;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.TodoComments
+{
+    internal class VisualStudioTodoCommentsService
+        : ForegroundThreadAffinitizedObject, ITodoCommentsService, ITodoCommentsServiceCallback, ITodoListProvider
+    {
+        private readonly VisualStudioWorkspaceImpl _workspace;
+        private readonly EventListenerTracker<ITodoListProvider> _eventListenerTracker;
+
+        private readonly ConcurrentDictionary<DocumentId, ImmutableArray<TodoCommentInfo>> _documentToInfos
+            = new ConcurrentDictionary<DocumentId, ImmutableArray<TodoCommentInfo>>();
+
+        /// <summary>
+        /// Our connections to the remote OOP server. Created on demand when we startup and then
+        /// kept around for the lifetime of this service.
+        /// </summary>
+        private KeepAliveSession? _keepAliveSession;
+
+        /// <summary>
+        /// Queue where we enqueue the information we get from OOP to process in batch in the future.
+        /// </summary>
+        private AsyncBatchingWorkQueue<DocumentAndComments> _workQueue = null!;
```

Comment:
>_workQueue [](start = 60, length = 10)

The same feedback as here: [https://github.com/dotnet/roslyn/pull/42383#pullrequestreview-375457477](https://github.com/dotnet/roslyn/pull/42383#pullrequestreview-375457477)